### ISSUE TITLE
Aborted (core dumped) training/test, cpu/gpu

### ISSUE THREAD

**Issue by ArashHosseini** [SeanNaren/deepspeech.pytorch#197]
Hey guys,
#104 @SeanNaren 
following issue by testing/eval the pre-trained model on cpu, also train on gpu
**machine 1**
**testing cpu**
ubuntu14.04,py3.5.pytorch0.3.0

```
~/git/deepspeech.pytorch$ python3.5 transcribe.py --model_path models/an4_pretrained.pth --audio_path /home/flyn/git/deepspeech.pytorch/sound/hello_world_exp.wav
{"_meta": {"acoustic_model": {"rnn_type": "gru", "name": "an4_pretrained.pth", "hidden_size": 800, "hidden_layers": 5, "version": "0.0.1"}, "decoder": {"type": "greedy", "alpha": null, "lm": false, "beta": null}, "language_model": {"name": null}}, "output": [{"transcription": "ROO O RA"}]}
*** Error in `python3.5': double free or corruption (!prev): 0x0000000001ff97e0 ***
Aborted (core dumped)
```
the same with training 

```
~/git/deepspeech.pytorch$ python3.5 train.py --train_manifest data/an4_train_manifest.csv --val_manifest data/an4_val_manifest.csv
Model Save directory already exists.
DeepSpeech(
  (conv): Sequential(
    (0): Conv2d (1, 32, kernel_size=(41, 11), stride=(2, 2), padding=(0, 10))
    (1): BatchNorm2d(32, eps=1e-05, momentum=0.1, affine=True)
    (2): Hardtanh(min_val=0, max_val=20, inplace)
    (3): Conv2d (32, 32, kernel_size=(21, 11), stride=(2, 1))
    (4): BatchNorm2d(32, eps=1e-05, momentum=0.1, affine=True)
    (5): Hardtanh(min_val=0, max_val=20, inplace)
  )
  (rnns): Sequential(
    (0): BatchRNN(
      (rnn): GRU(672, 800, bias=False, bidirectional=True)
    )
    (1): BatchRNN(
      (batch_norm): SequenceWise (
      BatchNorm1d(800, eps=1e-05, momentum=0.1, affine=True))
      (rnn): GRU(800, 800, bias=False, bidirectional=True)
    )
    (2): BatchRNN(
      (batch_norm): SequenceWise (
      BatchNorm1d(800, eps=1e-05, momentum=0.1, affine=True))
      (rnn): GRU(800, 800, bias=False, bidirectional=True)
    )
    (3): BatchRNN(
      (batch_norm): SequenceWise (
      BatchNorm1d(800, eps=1e-05, momentum=0.1, affine=True))
      (rnn): GRU(800, 800, bias=False, bidirectional=True)
    )
    (4): BatchRNN(
      (batch_norm): SequenceWise (
      BatchNorm1d(800, eps=1e-05, momentum=0.1, affine=True))
      (rnn): GRU(800, 800, bias=False, bidirectional=True)
    )
  )
  (fc): Sequential(
    (0): SequenceWise (
    Sequential(
      (0): BatchNorm1d(800, eps=1e-05, momentum=0.1, affine=True)
      (1): Linear(in_features=800, out_features=29)
    ))
  )
  (inference_softmax): InferenceBatchSoftmax(
  )
)
Number of parameters: 38067968
Epoch: [1][1/48]	Time 3.231 (3.231)	Data 0.127 (0.127)	Loss 103.3395 (103.3395)	
*** Error in `python3.5': double free or corruption (!prev): 0x0000000000e389a0 ***
Aborted (core dumped)
```
**same machine with gpu**
**ubuntu14.04,cuda8, cudnn6, py3.5.pytorch0.3.0**
```
Epoch: [70][46/48]	Time 0.172 (0.214)	Data 0.005 (0.011)	Loss 0.1023 (7.8711)	
Epoch: [70][47/48]	Time 0.203 (0.214)	Data 0.003 (0.011)	Loss 0.3170 (7.8689)	
Epoch: [70][48/48]	Time 0.256 (0.214)	Data 0.003 (0.011)	Loss 0.1704 (7.8665)	
Training Summary Epoch: [70]	Average Loss 0.211	
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7/7 [00:00<00:00,  7.32it/s]
Validation Summary Epoch: [70]	Average WER 25.535	Average CER 10.291	
Learning rate annealed to: 0.000000
Shuffling batches...
*** Error in `python3.5': free(): invalid pointer: 0x0000000001b52d80 ***
Aborted (core dumped)
```

**Machine 2**
**now i run the training at home machine, training is running until end then**
**ubuntu14.04,cuda9, cudnn6, py3.5.pytorch0.3.0**

```
python3.5 train.py --train_manifest data/an4_train_manifest.csv --val_manifest data/an4_val_manifest.csv --cuda
...
Epoch: [70][46/48]	Time 0.331 (0.205)	Data 0.003 (0.010)	Loss 0.2211 (7.0090)	
Epoch: [70][47/48]	Time 0.078 (0.205)	Data 0.002 (0.010)	Loss 0.0631 (7.0069)	
Epoch: [70][48/48]	Time 0.193 (0.205)	Data 0.002 (0.010)	Loss 0.1807 (7.0049)	
Training Summary Epoch: [70]	Average Loss 0.148	
100%|█████████████████████████████████████████████| 7/7 [00:00<00:00,  7.50it/s]
Validation Summary Epoch: [70]	Average WER 23.729	Average CER 9.222	
Learning rate annealed to: 0.000000
Shuffling batches...
*** Error in `python3.5': corrupted size vs. prev_size: 0x00000000020ca1b0 ***
Aborted (core dumped)
```
can somebody help please

**Comment by ArashHosseini** [comment]
```
sudo apt-get install libtcmalloc-minimal4
export LD_PRELOAD="/usr/lib/libtcmalloc_minimal.so.4"
```
solved my issue

**Comment by alfredcs** [comment]
Setting LD_PRELOAD="/usr/lib/libtcmalloc_minimal.so.4" didn't help for me.
I am on CentOS7 with Nvidia 8xV100 on cuda9.x. Thanks in advance!

$ python train.py --cuda --train-manifest data/an4_train_manifest.csv --val-manifest data/an4_val_manifest.csv 
DeepSpeech(
  (conv): MaskConv(
    (seq_module): Sequential(
      (0): Conv2d(1, 32, kernel_size=(41, 11), stride=(2, 2), padding=(20, 5))
      (1): BatchNorm2d(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (2): Hardtanh(min_val=0, max_val=20, inplace)
      (3): Conv2d(32, 32, kernel_size=(21, 11), stride=(2, 1), padding=(10, 5))
      (4): BatchNorm2d(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (5): Hardtanh(min_val=0, max_val=20, inplace)
    )
  )
  (rnns): Sequential(
    (0): BatchRNN(
      (rnn): GRU(1312, 800, bidirectional=True)
    )
    (1): BatchRNN(
      (batch_norm): SequenceWise (
      BatchNorm1d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True))
      (rnn): GRU(800, 800, bidirectional=True)
    )
    (2): BatchRNN(
      (batch_norm): SequenceWise (
      BatchNorm1d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True))
      (rnn): GRU(800, 800, bidirectional=True)
    )
    (3): BatchRNN(
      (batch_norm): SequenceWise (
      BatchNorm1d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True))
      (rnn): GRU(800, 800, bidirectional=True)
    )
    (4): BatchRNN(
      (batch_norm): SequenceWise (
      BatchNorm1d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True))
      (rnn): GRU(800, 800, bidirectional=True)
    )
  )
  (fc): Sequential(
    (0): SequenceWise (
    Sequential(
      (0): BatchNorm1d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (1): Linear(in_features=800, out_features=29, bias=False)
    ))
  )
  (inference_softmax): InferenceBatchSoftmax()
)
Number of parameters: 41187968
Segmentation fault (core dumped)

**Comment by alfredcs** [comment]
> Setting LD_PRELOAD="/usr/lib/libtcmalloc_minimal.so.4" didn't help for me.
> I am on CentOS7 with Nvidia 8xV100 on cuda9.x. Thanks in advance!
> 
> $ python train.py --cuda --train-manifest data/an4_train_manifest.csv --val-manifest data/an4_val_manifest.csv
> DeepSpeech(
> (conv): MaskConv(
> (seq_module): Sequential(
> (0): Conv2d(1, 32, kernel_size=(41, 11), stride=(2, 2), padding=(20, 5))
> (1): BatchNorm2d(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
> (2): Hardtanh(min_val=0, max_val=20, inplace)
> (3): Conv2d(32, 32, kernel_size=(21, 11), stride=(2, 1), padding=(10, 5))
> (4): BatchNorm2d(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
> (5): Hardtanh(min_val=0, max_val=20, inplace)
> )
> )
> (rnns): Sequential(
> (0): BatchRNN(
> (rnn): GRU(1312, 800, bidirectional=True)
> )
> (1): BatchRNN(
> (batch_norm): SequenceWise (
> BatchNorm1d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True))
> (rnn): GRU(800, 800, bidirectional=True)
> )
> (2): BatchRNN(
> (batch_norm): SequenceWise (
> BatchNorm1d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True))
> (rnn): GRU(800, 800, bidirectional=True)
> )
> (3): BatchRNN(
> (batch_norm): SequenceWise (
> BatchNorm1d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True))
> (rnn): GRU(800, 800, bidirectional=True)
> )
> (4): BatchRNN(
> (batch_norm): SequenceWise (
> BatchNorm1d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True))
> (rnn): GRU(800, 800, bidirectional=True)
> )
> )
> (fc): Sequential(
> (0): SequenceWise (
> Sequential(
> (0): BatchNorm1d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
> (1): Linear(in_features=800, out_features=29, bias=False)
> ))
> )
> (inference_softmax): InferenceBatchSoftmax()
> )
> Number of parameters: 41187968
> Segmentation fault (core dumped)

The core dump was caused by this. Upgraded gcc to 5.x seemed to help. 
......
Program received signal SIGSEGV, Segmentation fault.
0x00007fffa41c04fc in construct<_object*, _object*> (__p=0xb, this=0x555556782468)
    at ~/anaconda3/gcc/include/c++/ext/new_allocator.h:120
120		{ ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
Missing separate debuginfos, use: debuginfo-install glibc-2.17-260.el7_6.3.x86_64 gsm-1.0.13-11.el7.x86_64 libpng-1.5.13-7.el7_2.x86_64 sox-14.4.1-6.el7.x86_64
